File: Python/data/binTree.py
class bin_tree( object ):
	def __init__(self):
		self.d = None
		self.l = None
		self.r = None
	def add(self, n, nbit, v):
		p = 1 << (nbit - 1)
		t = self
		while(p != 0):
			c = n & p
			p >>= 1
			if c == 0:
				if t.l == None:
					t.l = bin_tree()
				t = t.l
			else:
				if t.r == None:
					t.r = bin_tree()
				t = t.r
		t.d = v
	def find(self, br):
		t = self
		while(True):
			b = br.read1()
			if b == None:
				break
			if b == 0:
				t = t.l
			else:
				t = t.r
			if t == None:
				break
			if t.d != None:
				return t.d
		return None

File: JavaScript/data/bin_tree.js
class binTree {
	constructor(){
		this.l = null;
		this.r = null;
		this.val = null;
	}
	add(n, nbit, v){
		let f = 1 << (nbit - 1);
		let e = this;
		while(f != 0){
			const c = n & f;
			if(c == 0){
				if(e.l == null){
					e.l = new binTree();
				}
				e = e.l;
			}
			else{
				if(e.r == null){
					e.r = new binTree();
				}
				e = e.r;
			}
			f >>= 1;
		}
		e.val = v;
	}
	find(br){
		let e = this;
		for(;;){
			const b = br.read1();
			e = b == 0 ? e.l : e.r;
			if(e == null)
				return null;
			if(e.val != null)
				return e.val;
		}
	}
}
